File: 0101-0200/0166.Fraction_to_Recurring_Decimal.java
//1ms





public class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) {
            return "0";
        }
        StringBuilder res = new StringBuilder();
        // "+" or "-"
        res.append(((numerator > 0) ^ (denominator > 0)) ? "-" : "");
        long num = Math.abs((long)numerator);
        long den = Math.abs((long)denominator);
        
        // integral part
        res.append(num / den);
        num %= den;
        if (num == 0) {
            return res.toString();
        }
        
        // fractional part
        res.append(".");
        HashMap<Long, Integer> map = new HashMap<Long, Integer>();
        map.put(num, res.length());
        while (num != 0) {
            num *= 10;
            res.append(num / den);
            num %= den;
            if (map.containsKey(num)) {
                int index = map.get(num);
                res.insert(index, "(");
                res.append(")");
                break;
            }
            else {
                map.put(num, res.length());
            }
        }
        return res.toString();
    }
}





//0ms






class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        // Step 1: Convert numerator and denominator to long to avoid overflow
        long lnum = numerator, lden = denominator;

        // Step 2: Handle the zero case
        if (lnum == 0) {
            return "0";
        }

        // Step 3: If the result divides evenly (no decimal part), return the integer part
        if (lnum % lden == 0) {
            return Long.toString(lnum / lden);
        }

        // Step 4: Create a StringBuilder to construct the result
        StringBuilder number = new StringBuilder();

        // Step 5: Handle the negative sign
        // XOR operator: if only one is negative, then result is negative
        if ((lden < 0) ^ (lnum < 0)) {
            number.append('-');
        }

        // Step 6: Convert both numbers to positive
        lnum = Math.abs(lnum);
        lden = Math.abs(lden);

        // Step 7: Append the integer part of the division
        number.append(lnum / lden);
        number.append('.');  // Add the decimal point

        // Step 8: Reduce the fraction by their GCD to simplify
        long GCD = gcd(lnum, lden);
        lnum /= GCD;
        lden /= GCD;

        // Step 9: Begin processing the fractional part
        lnum %= lden;   // Get initial remainder
        lnum *= 10;     // Start long division

        // Step 10: Check if the denominator has only 2s and 5s as prime factors
        // If so, it is a terminating decimal
        int twos = 0, fives = 0;
        long temp = lden;

        // Count how many times denominator is divisible by 2
        while (temp % 2 == 0) {
            twos++;
            temp /= 2;
        }

        // Count how many times denominator is divisible by 5
        while (temp % 5 == 0) {
            fives++;
            temp /= 5;
        }

        // Step 11: If only 2s and 5s, it's a terminating decimal
        if (temp == 1) {
            // Keep dividing and appending until there's no remainder
            while (lnum != 0) {
                number.append(lnum / lden);
                lnum = (lnum % lden) * 10;
            }
        } else {
            // Step 12: If not terminating, we might have a non-repeating prefix
            int nonRepLen = Math.max(twos, fives);

            for (int i = 0; i < nonRepLen; i++) {
                number.append(lnum / lden);
                lnum = (lnum % lden) * 10;
            }

            // Step 13: Detect repeating cycle by tracking remainders
            long initRem = lnum;  // Save initial remainder
            StringBuilder repeating = new StringBuilder();

            do {
                repeating.append(lnum / lden);
                lnum = (lnum % lden) * 10;
            } while (lnum != initRem);  // Stop when remainder repeats

            // Step 14: Wrap repeating part in parentheses
            number.append('(').append(repeating.toString()).append(')');
        }

        // Step 15: Return final string
        return number.toString();
    }

    // Helper method to compute GCD (Greatest Common Divisor)
    private long gcd(long a, long b) {
        return (b == 0) ? a : gcd(b, a % b);
    }
}

File: 0101-0200/0166.Fraction_to_Recurring_Decimal.js
//0ms






/**
 * @param {number} numerator
 * @param {number} denominator
 * @return {string}
 */
var fractionToDecimal = function(numerator, denominator) {
    if (numerator === 0) return "0";
    let res = "";
    if (Math.sign(numerator) !== Math.sign(denominator)) res += "-";

    let num = Math.abs(numerator);
    let den = Math.abs(denominator);
    res += Math.floor(num / den);
    num %= den;
    if (num === 0) return res;

    res += ".";
    const map = new Map();

    while (num !== 0) {
        if (map.has(num)) {
            const idx = map.get(num);
            return res.slice(0, idx) + "(" + res.slice(idx) + ")";
        }
        map.set(num, res.length);
        num *= 10;
        res += Math.floor(num / den);
        num %= den;
    }

    return res;
};






//1ms






/**
 * @param {number} numerator
 * @param {number} denominator
 * @return {string}
 */
var fractionToDecimal = function(numerator, denominator) {
    if (numerator === 0) {
        return "0";
    }

    const isNegative = (numerator < 0) !== (denominator < 0);
    
    numerator = Math.abs(numerator);
    denominator = Math.abs(denominator);

    let digit = Math.floor(numerator / denominator);
    let result = digit.toString();
    let decimal = [];
    let remainder = numerator % denominator;
    let count = 10000;
    let decimalMap = new Map();
    let i = 0;
    let decimalRepeatIndex = -1;

    numerator = remainder * 10;

    while (numerator) {
        digit = Math.floor(numerator / denominator)
        remainder = numerator % denominator;
        const mapKey = `${remainder},${digit}`;

        if (decimalMap.has(mapKey)) {
            decimalRepeatIndex = decimalMap.get(mapKey);

            break;
        }

        decimal.push(digit);
        numerator = remainder * 10;
        decimalMap.set(mapKey, i);
        i++;

        if (!count--) {
            break;
        }
    }
    
    if (decimal.length) {
        result += ".";
        const index = decimalRepeatIndex !== -1 ? decimalRepeatIndex : decimal.length;
        let repeatingDecimals = "";

        for (let j = 0; j < index; j++) {
            result += decimal[j];
        }

        for (let j = index; j < decimal.length; j++) {
            repeatingDecimals += decimal[j];
        }

        if (repeatingDecimals) {
            result += "(" + repeatingDecimals + ")";
        }
    } 

    return isNegative ? "-" + result : result;
};






//1ms







/**
 * @param {number} numerator
 * @param {number} denominator
 * @return {string}
 */
var fractionToDecimal = function(numerator, denominator) {
    // check for repeating digits
    const map = new Map();

    let result = '';

    const sign = Math.sign(numerator) * Math.sign(denominator)
    if (sign < 0) result += '-';

    const den = Math.abs(denominator);
    let num = Math.abs(numerator);

    // get digits before decimal
    result += Math.floor(num / den);
    num %= den;
    if (num > 0) result += '.';
    num *= 10;

    // get digits after decimal + check for repeat
    while (num > 0) {
        let digit = Math.floor(num / den);
        if (map.has(num)) {
            const idx = map.get(num);
            result = `${result.slice(0, idx)}(${result.slice(idx)})`;
            num = 0;
        } else {
            map.set(num, result.length);
            result += digit;
            num %= den;
            num *= 10;
        }
    }
    return result;
};

